### PR TITLE
Extract queue_job module from connector

### PR DESCRIPTION
**Pull request by guewen**
* Replace ConnectorSession by odoo.api.Environment
* Rework job classes
* Jobs are now applied on Model methods

```python
  class MyModel(models.Model):
     _name = 'my.model'

     @api.multi
     @job
     def my_method(self, a, k=None):
         _logger.info('executed with a: %s and k: %s', a, k)


  class MyOtherModel(models.Model):
      _name = 'my.other.model'

      @api.multi
      def button_do_stuff(self):
          self.env['my.model'].with_delay().my_method('a', k=2)
```

See history on https://github.com/OCA/connector/pull/212

### CODE REVIEW

File: queue_job/__manifest__.py
Comment (lmignon) [inline queue_job/__manifest__.py]: Is-it really an application?

**Comment by lmignon** [comment]
@guewen What a great refactoring!!!!! No time for a full review yet. After a first read I have not found anything to complain. I need to investigate a little further to completly understand what is behind the changes and for exemple at https://github.com/OCA/queue/pull/1/files#diff-8b63facccbd0e181392cdc06745d034eR615

File: queue_job/__manifest__.py
Comment (guewen) [reply queue_job/__manifest__.py]: Isn't it an application if we have a top-level menu for accessing the queue models (Channels, Job functions and Jobs). Corollary, would we want a top-level menu? (I'd say yes, and we need to find an icon BTW).

**Comment by guewen** [comment]
I still have to test in _in real_ with `base_import_async` because it's only tested with the python tests for now :)

**Comment by leorochael** [comment]
As a test, I've tried to migrate a little project that doesn't try to use the whole connector to use just this PR, but quickly ran into an issue: We also use the connector events subsystem...

It seems to me that many uses of the queue without the connector will use (or could benefit from the use of) the events as well.

Should we try to move it along into `queue` or should it perhaps have its own repo?

**Comment by leorochael** [comment]
Speaking of moving modules into the queue repo, I think base_import_async would be a good candidate, as it uses solely the job queue.

Rewriting base_import_async to use the new model method delay would be a good way to taste what that api should feel like.

**Comment by leorochael** [comment]
And speaking of model methods delay, if we add a method to the global BaseModel object that allows delaying methods, why not simply allow delaying all methods (with serializable parameters) by default?

That's what ERP5 does, by the way.

You could then just add decorators to the methods where you want to pass different queue parameters than the defaults (like priority, or channel name).

**Comment by guewen** [comment]
> It seems to me that many uses of the queue without the connector will use (or could benefit from the use of) the events as well.

You might be right. However, it might be done in a subsequent pull request.

**Comment by guewen** [comment]
> why not simply allow delaying all methods (with serializable parameters) by default?

I thought about that. Actually, I am _restricting_ to delay methods with the `@job` decorator so it would just need to remove this restriction. 

We should just think about the security. The `with_delay` does a `getattr` on the model, so I used the `@job` decorator as a whitelist of the jobs to run. Before removing the restriction, we must be sure that we don't open a security hole.

The other point is that currently we show the "job functions" in a view so we can configure the channels. It might be weird to have only some of them, but maybe that's just a matter of improving the way we present and name them.

**Comment by lmignon** [comment]
> It seems to me that many uses of the queue without the connector will use (or could benefit from the use of) the events as well.

IMO the events subsystem is also a good candidate for a generic module not related at all to the connector module nor the queue_job module.

**Comment by leorochael** [comment]
@lmignon I agree, but even as an independent module, should it be in a separate repo?

**Comment by lmignon** [comment]
@leorochael IMO yes it should be in a separate repo with its own release management/cycle.

File: queue_job/__manifest__.py
Comment (blaggacao) [reply queue_job/__manifest__.py]: Although they are, I would argue they shouldn't. I feel coining it "application" should account mainly for the business end user perspective. So, If I can't magically hire the best people in the world with this or similar stuff, then it's probably not an "application". This implies abstaining from top-level menu. Maybe it's better classified (hidden) in the "technical" settings menu.

Comment (guewen) [reply queue_job/__manifest__.py]: @blaggacao if the user has not the accesses rights for the queue, he won't see the top-level menu, so for a technical end-user, it can make sense to have a top-level menu (jobs is a menu you open often as you regularly want to check them, it should not be hidden in an obscure place).

Comment (lmignon) [reply queue_job/__manifest__.py]: @guewen your POV makes sense :+1:

Comment (leorochael) [reply queue_job/__manifest__.py]: Even if it does have a top level menu, I'm not sure it's enough to call it an application.

Personally, I'd be ok if the Queue menu was somewhere inside Technical instead of being top-level.

If I need to keep an eye on the jobs queue all the time so that a top-level menu makes sense, then I'd argue something is wrong in my application...

Comment (guewen) [reply queue_job/__manifest__.py]: Then you never tried to synchronize with Magento :joy: .
Joke apart, I'm not sure what exactly is supposed to be an Application. So given the controversy, I removed the flag.

Comment (blaggacao) [reply queue_job/__manifest__.py]: Convincing and true. I was wrong.

**Comment by adrienpeiffer** [comment]
Hi @guewen,

Thanks for the work !
I did some functional test on this one. I just created a PR (https://github.com/guewen/queue/pull/2) on your own branch to fix some things that appeared during my tests.

We can discuss on each of my commits if you want.

Thanks,

Adrien

**Comment by guewen** [comment]
@adrienpeiffer @lmignon @sbidoul @leorochael @blaggacao  The last 2 commits rework the final bits of the jobs and related actions implementation. Tests are green on my computer. Waiting for the Travis build. Hopefully merged in 2016? :-)

**Comment by guewen** [comment]
Aaaand it's green!

**Comment by parthivgls** [comment]
👍 +1

File: queue_job/controllers/main.py
Comment (sbidoul) [inline queue_job/controllers/main.py:62]: Don't we need some kind of sudo(job.user_id) here?

File: queue_job/controllers/main.py
Comment (sbidoul) [inline queue_job/controllers/main.py]: job.env.uid is always SUPERUSER_ID

File: queue_job/__manifest__.py
Comment (sbidoul) [inline queue_job/__manifest__.py]: s,Acsone,ACSONE SA/NV,

File: queue_job/models/base.py
Comment (sbidoul) [inline queue_job/models/base.py:23]: @guewen nice!

File: queue_job/controllers/main.py
Comment (guewen) [reply queue_job/controllers/main.py:62]: the sudo is done in `perform()` that executes `self.func()`, the latter being the method call on the sudoed recordset: https://github.com/guewen/queue/blob/43bb875adc421f6773a07ae6ba143014d7a0e600/queue_job/job.py#L419

File: queue_job/controllers/main.py
Comment (guewen) [reply queue_job/controllers/main.py]: Do you mean we should replace `job.env.uid` by `SUPERUSER_ID`? I replaced by `job.env(cr=new_cr)` which returns the new env with only the `cr` modified, avoiding some repetition.

File: queue_job/__manifest__.py
Comment (guewen) [reply queue_job/__manifest__.py]: Fixed

**Review by sbidoul — APPROVED** [review]
Great job!

**Comment by leorochael** [comment]
Hi @guewen, one thing I'm missing in this PR is the configuration file effort I did for `connector` on OCA/connector#213.

Since there wasn't a maintained 10.0 branch of connector yet, I ended up not porting it over before your queue_job extraction effort, even though it was already landed on 9.0 and it would obviously be trivial to make it work on 10.0.

That PR, along with odoo/odoo#13685, which got squash merged on odoo/odoo@5ee4344, means it should be possible to configure and launch a `queue_job` enabled odoo 10.0 entirely from `odoo.cfg` without having to resort to environment variables.

I'd like to see the config file work land on queue_job before this PR is merged, but I'm not sure when I'm going to find the [round tuits](https://en.wiktionary.org/wiki/round_tuit) to make it happen on my side, and I don't want to hold this PR hostage, so feel free to land it without.

The worse that could happen is that I'll make another PR later...

**Comment by adrienpeiffer** [comment]
Hi Guewen,
Can you take a look on https://github.com/guewen/queue/pull/3 ?

**Comment by adrienpeiffer** [comment]
Hi,

One last thing, I noticed that when an object in memory that is in parameter of the delayed method, and that this object is modified (like a dictionary). When the job fails, changes made to the dictionary remain persistent. Any comment ?

File: queue_job/README.rst
Comment (pedrobaeza) [inline queue_job/README.rst]: Put this information

File: queue_job/README.rst
Comment (pedrobaeza) [inline queue_job/README.rst]: Remove this

**Comment by guewen** [comment]
@adrienpeiffer could you give an example?

**Comment by leorochael** [comment]
@adrienpeiffer, this is normal behaviour in Python,  if you pass mutable objects into a function/method and the function mutates the object, the object remains mutated outside of the scope of the function, even if the function raises an exception.

Only ORM records know to revert state when a transaction is aborted.

The solution is either not to mutate the object in the delayed call, or to make a copy if you need to.

**Comment by leorochael** [comment]
@guewen: guewen/queue#5 for the config file work.

**Comment by adrienpeiffer** [comment]
@leorochael Of course. But in this case, these mutable objects are stored in the database when the job failed. So, I think that it's not really clean.

@guewen. For instance, I give a dictionary that contains all data necessary the create a new partner. When the job failed (for any reason) the dictionary (args, kwargs) of the delayed method contains other key 
in the dictionary like ```message_followers_ids``` or something like that.

Personally, I just make a deep copy of the dictionary in the delayed method when the job is running.

**Comment by guewen** [comment]
@adrienpeiffer Nice catch! Thanks for the explanation, I could fix it in efddad9.

@leorochael Thanks for bringing this feature back. I added a comment on the pull request regarding the naming of the section `options-queue_job`.

**Comment by sbidoul** [comment]
And when https://github.com/guewen/queue/pull/5 is merged, I'll do a PR for the sequential channels feature.

**Review by adrienpeiffer — APPROVED** [review]
:+1: Great work !

**Comment by sbidoul** [comment]
@guewen here is the last piece of the puzzle: https://github.com/guewen/queue/pull/6

**Comment by guewen** [comment]
Thanks @sbidoul. I'm eager to see this merged :)

**Comment by sbidoul** [comment]
Voilà @guewen. À toi l'honneur du tweet.

**Comment by blaggacao** [comment]
@guewen Congratulations! This is actually really exciting work. To give people like me a headstart, consider a possible use cases section in the readme? Like advertising it a little bit, in the words of a seasoned contributor?

**Comment by leorochael** [comment]
We should consider adding a github-pages or ReadTheDocs auto-generated site next.

Also, we should probably port [base_import_async](https://github.com/OCA/connector-interfaces/tree/8.0/base_import_async) (and perhaps move it into this repo) as an example of what can be done with `queue_job`.

**Pull request merged**
